File: src/main/java/cn/chahuyun/economy/entity/UserBackpack.java
package cn.chahuyun.economy.entity;

import cn.chahuyun.economy.prop.PropBase;
import cn.chahuyun.economy.prop.PropsManager;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.io.Serializable;

/**
 * 用户背包
 *
 * @author Moyuyanli
 * @date 2022/11/15 9:02
 */
@Entity(name = "UserBackpack")
@Table
@Getter
@Setter
public class UserBackpack implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    /**
     * 用户id
     */
    private String userId;
    /**
     * 道具编码
     */
    private String propCode;
    /**
     * 道具类型
     */
    private String propKind;
    /**
     * 道具id
     */
    private Long propId;

    //    @ManyToOne
//    private UserInfo userInfo;
    public UserBackpack() {
    }

    public UserBackpack(String userId, String propCode, String propKind, Long propId) {
        this.userId = userId;
        this.propCode = propCode;
        this.propKind = propKind;
        this.propId = propId;
    }

    public Long getPropId() {
        return propId;
    }

    public String getUserId() {
        return userId;
    }

    /**
     * 获取改背包道具
     *
     * @param tClass 类型
     * @return 道具
     */
    public <T extends PropBase> T getProp(Class<T> tClass) {
        return PropsManager.getProp(this, tClass);
    }

}

File: src/main/kotlin/cn/chahuyun/economy/entity/UserBackpack.kt
package cn.chahuyun.economy.entity

import cn.chahuyun.economy.prop.BaseProp
import cn.chahuyun.economy.prop.PropsManager
import jakarta.persistence.*
import java.io.Serializable

/**
 * 用户背包
 *
 * @author Moyuyanli
 * @date 2022/11/15 9:02
 */
@Entity(name = "UserBackpack")
@Table
class UserBackpack(
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    val id: Long? = null,

    /**
     * 用户id
     */
    val userId: String? = null,

    /**
     * 道具编码
     */
    val propCode: String? = null,

    /**
     * 道具类型
     */
    val propKind: String? = null,

    /**
     * 道具id
     */
    val propId: Long? = null
) : Serializable {

    /**
     * 获取该背包道具
     *
     * @return 道具
     */
    inline fun <reified T : BaseProp> getProp(): T {
        return PropsManager.getProp(this, T::class.java)
    }

    /**
     * 获取指定类型的属性对象
     *
     * @param T 泛型类型参数，必须继承自BaseProp
     * @param tClass 要获取的属性类型的Class对象
     * @return 返回指定类型的属性实例
     */
    fun <T : BaseProp> getProp(tClass: Class<T>): T {
        return PropsManager.getProp(this, tClass)
    }

    /**
     * 获取该背包道具 (兼容旧版 API)
     */
    @Deprecated("使用泛型版本 getProp(clazz)", ReplaceWith("getProp()"))
    fun getPropLegacy(): BaseProp? {
        return PropsManager.getProp(this)
    }
}
